feat(core): add comprehensive transaction propagation support

junhyeongkim2 opened the pull request:
## Summary

Implements comprehensive transaction propagation support for MikroORM, providing fine-grained control over transaction boundaries

  ## What's Changed

  - Added `TransactionManager` class to handle transaction lifecycle and propagation
  - Implemented all 7 standard propagation types:
    - `REQUIRED`: Join existing transaction or create new one
    - `REQUIRES_NEW`: Always create independent transaction
    - `NESTED`: Create savepoint if transaction exists, otherwise new transaction
    - `SUPPORTS`: Use transaction if available, otherwise execute non-transactionally
    - `MANDATORY`: Require existing transaction, throw error if none exists
    - `NEVER`: Must execute without transaction, throw error if one exists
    - `NOT_SUPPORTED`: Suspend existing transaction and execute without transaction
  - Added comprehensive documentation with examples for each propagation type
  - Full test coverage for all propagation scenarios
  - TypeScript-safe implementation with proper error handling

  ## Breaking Changes

  None - fully backward compatible. The default behavior remains unchanged.

  Fixes #6788

junhyeongkim2 commented on packages/core/src/index.ts:
```diff
@@ -32,3 +32,4 @@ export * from './naming-strategy';
 export * from './metadata';
 export * from './cache';
 export * from './decorators';
+export * from './transaction';
```
Moved to utils/ as suggested.

B4nan commented:
> 📝 Step 2: MongoDB Support
> 📝 Step 3: SQLite Support

What exactly is needed for those? The changes are already implemented on the core level, so they should just work with all drivers, no? Would be good to test it with those drivers too.

> 📝 Step 4: Platform Adapter Integration

I don't understand what this should be about.

> 📝 Step 5: Documentation & Examples

This is a requirement for merging any PR. Ideally, I'd like to see all the steps implemented here. If you want to split into multiple PRs, you can do so, but I won't be merging incomplete PRs to master - ideally this would end up as a single commit in there.

junhyeongkim2 commented:
@B4nan  Added tests and documentation as requested

  - **Tests**: All database drivers covered. Note: SQLite excluded (60s timeout due to single-writer limitation), MongoDB validates "Transaction already in progress" error for NESTED.

  - **Docs**: Minimal propagation section with table and example added to avoid duplication.

  Please take a look and let me know if any adjustments needed. Thanks!

junhyeongkim2 commented:
1. Migrated transaction propagation tests to /tests/features/transactions/ with direct MikroORM.init() calls
  2. Unified PostgreSQL/MySQL/MariaDB/MSSQL tests using describe.each pattern (47 tests × 4 DBs)
  3. Separated SQLite and MongoDB due to unique transaction limitations
  4. Skipped 7 MSSQL tests due to database engine limitations

  @B4nan Would appreciate your review when convenient. Thanks!

B4nan commented:
I am looking at the other propagation options described in #6788, and it feels like they should be pretty trivial to implement. If that's the case, can you give it a try?

B4nan commented on docs/docs/transactions.md:
```diff
@@ -83,6 +83,32 @@ Explicit transaction demarcation is required when you want to include custom DBA
 
+#### Transaction Propagation
+
+Control how nested transactions behave using the `propagation` option:
```
the sprint docs could be used as an inspiration, we dont need to be that verbose, but what we have now is way too little

https://www.baeldung.com/spring-transactional-propagation-isolation

junhyeongkim2 commented:
Hi @B4nan, thanks for the review! Here's what I've addressed:

  1. Documentation expanded ✅

  - Added comprehensive explanations following Spring docs style and existing MikroORM patterns
  - Each type now includes description, use case, and examples

  2. Additional propagations implemented ✅

  - SUPPORTS, MANDATORY, NEVER added as suggested

  3. SUPPORTS, MANDATORY, NEVER Tests added ✅

  - 36 test cases covering all scenarios
  - Works across all databases

  Let me know if you'd like any adjustments!

junhyeongkim2 commented on packages/core/src/utils/TransactionManager.ts:
```diff
@@ -0,0 +1,342 @@
+  ): Promise<T> {
+    const em = this.em.getContext(false) as EntityManager;
+
+    // If no explicit propagation is set, use default behavior
+    if (!options.propagation) {
+      return this.executeDefaultTransaction(em, cb, options);
+    }
```
You're right - the default behavior and NESTED are functionally identical. 

  We could eliminate this duplication by setting NESTED as the explicit default:
  options.propagation ??= TransactionPropagation.NESTED;

  Then remove executeDefaultTransaction entirely. What do you think?

B4nan replied:
Yes please, also update the docs about this. It feels really weird to have some default mode that is not one of the existing options, and clearly this is a code duplication too.

junhyeongkim2 commented on packages/core/src/utils/TransactionManager.ts:
```diff
@@ -0,0 +1,342 @@
+    cb: (em: EntityManager) => T | Promise<T>,
+    options: TransactionOptions,
+  ): Promise<T> {
+    const fork = this.createFork(em, options);
```
You're right Simply returning cb(em) correctly represents the joining semantics without unnecessary overhead. Thanks for catching this!

junhyeongkim2 commented:
@B4nan Thanks for the review! I've made the following changes:

  1. **Removed `executeDefaultTransaction` and `processTransactionCallback`** - Set NESTED as explicit default instead
  2. **Simplified `joinExistingTransaction`** - Now just returns `cb(em)` without creating unnecessary fork/TransactionContext
  3. **Updated documentation** - Clarified that NESTED is the default propagation type

  All tests are passing. Let me know if you'd like any adjustments!

junhyeongkim2 commented on packages/core/src/utils/TransactionManager.ts:
```diff
@@ -0,0 +1,292 @@
+    cb: (em: EntityManager) => T | Promise<T>,
+    options: TransactionOptions = {},
+  ): Promise<T> {
+    const em = this.em.getContext(false) as EntityManager;
```
Agreed, removing the cast.

junhyeongkim2 commented:
@B4nan  Thanks for pointing out the docs ordering - fixed! Also removed the unnecessary type cast. Appreciate the detailed review.

B4nan commented on packages/core/src/errors.ts:
```diff
@@ -301,3 +301,19 @@ export class NotFoundError<T extends AnyEntity = AnyEntity> extends ValidationEr
 
 }
+
+export class TransactionStateError<T extends AnyEntity = AnyEntity> extends ValidationError<T> {
```
I'd remove the type argument, such errors are not really bound to entity type

```suggestion
export class TransactionStateError extends ValidationError {
```

B4nan reviewed (approved):
all right, we are good to go, thanks a lot for the PR!

injae-kim commented:
Aweeeeeesome work @junhyeongkim2 👍
Thanks for the kind review to reviewer.
